File: leetcode/401.java
/*
 * 401. Binary Watch
 */
public class Solution {
    private int index(int num) {
        int cnt = 0;
        for (int i = 0; i < 6; ++i) {
            if ((num & 1) == 1) cnt++;
            num >>= 1;
        }
        return cnt;
    }

    public List<String> readBinaryWatch(int num) {
        List<List<Integer>> hours = new ArrayList<>();
        List<List<Integer>> minutes = new ArrayList<>();
        for (int i = 0; i < 6; ++i) {
            hours.add(new ArrayList<>());
            minutes.add(new ArrayList<>());
        }

        for (int i = 0; i < 60; ++i) {
            if (i < 12) hours.get(index(i)).add(i);
            minutes.get(index(i)).add(i);
        }

        List<String> res = new ArrayList<>();
        for (int i = 0; i <= num; ++i) {
            int j = num - i;
            if (i >= hours.size()) break;
            if (j >= minutes.size()) continue;

            for (int x = 0; x < hours.get(i).size(); ++x) {
                for (int y = 0; y < minutes.get(j).size(); ++y) {
                    StringBuilder str = new StringBuilder();
                    str.append(hours.get(i).get(x));
                    str.append(":");
                    if (minutes.get(j).get(y) < 10) {
                        str.append("0");
                    }
                    str.append(minutes.get(j).get(y));
                    res.add(str.toString());
                }
            }
        }
        return res;
    }
}

File: leetcode/401.cpp
// 401. Binary Watch
#include <iostream>
#include <vector>
#include <string>
using namespace std;

class Solution {
public:
    vector<string> readBinaryWatch(int num) {
        vector<string> result;
        vector<int> hours{8, 4, 2, 1};
        vector<int> minutes{32, 16, 8, 4, 2, 1};

        for (int i = 0; i <= num; ++i) {
            vector<int> hour_num = generate(hours, i);
            vector<int> minute_num = generate(minutes, num - i);

            for (int j = 0; j < hour_num.size(); ++j) {
                if (hour_num[j] >= 12) continue;
                for (int k = 0; k < minute_num.size(); ++k) {
                    if (minute_num[k] >= 60) continue;
                    result.push_back(to_string(hour_num[j]) + ":" + 
                        (minute_num[k] < 10 ? "0" + to_string(minute_num[k]) : to_string(minute_num[k])));
                }
            }
        }
        
        return result;
    }

private:
    vector<int> generate(vector<int> nums, int count) {
        vector<int> result;
        generateHelper(nums, count, 0, 0, result);
        return result;
    }

    void generateHelper(vector<int> nums, int count, int pos, int sum, vector<int>& result) {
        if (count == 0) {
            result.push_back(sum);
            return;
        }

        for (int i = pos; i < nums.size(); ++i) {
            generateHelper(nums, count - 1, i + 1, sum + nums[i], result);
        }
    }
};

int main() {
    Solution s;
    vector<string> res = s.readBinaryWatch(1);
    for (string s : res) {
        cout << s << " ";
    }
    cout << "\n";
}